Make PropertyGraph type ids modifiable

scober opened the pull request:
Using the same approach for modifying the contents of the topology arrays in GraphTopology.

ddn0 commented:
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

I'm always surprised that lots of IDEs don't follow this convention.

witchel commented:
> https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline
> 
> I'm always surprised that lots of IDEs don't follow this convention.

If our own experience is any guide, I think it's just something that is easy to slip by.

Commit:
PropertyGraph type ids are modifiable

Using the same approach for modifying the contents of the topology
arrays in GraphTopology.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -413,6 +414,16 @@ katana::PropertyGraph::Make(katana::GraphTopology&& topo_to_assign) {
+    NUMAArray<EntityTypeID> edge_entity_type_id) {
+  return std::make_unique<katana::PropertyGraph>(
+      std::move(topo_to_assign), std::move(node_entity_type_id),
+      std::move(edge_entity_type_id));
```
How does this work without the entity-type-id maps?

scober replied:
I actually hadn't thought it out that far. I suppose if we are passing in these arrays at construction time, we should be passing in the maps at construction time as well.

roshandathathri replied:
Yes, otherwise, there's no way to decouple it. I'm working on moving the "maps" to a single class. But for now, you can pass each of those during construction.

scober replied:
And then we will have to do the "infer all maps from the minimal set of maps" step here as well.

roshandathathri replied:
No, the class object will do the inferral if needed.

scober replied:
I know I pushed for a "just construct the full set of maps from the minimal set" approach, but I am changing my tune. The minimal set of maps is 4 maps and the full set is 8. I think we should wait to pass maps here until we have a sensible class wrapping them.

If we want to generate some maps from others, we should put that logic in the class that wraps the maps (looks like Roshan beat me to this one).

amberhassaan replied:
Agree, PropertyGraph should keep a reference or copy of a `EntityTypeMappings` class that encapsulates the type ID arrays and maps (and the API).

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -413,6 +414,20 @@ katana::PropertyGraph::Make(katana::GraphTopology&& topo_to_assign) {
+// TODO (scober): This Make variant should also pass an EntityTypeManager (once
+// that class has been defined). The equivalent in the current code base would
+// be to pass all 8 of the PropertyGraph-internal maps that manipulate
+// EntityTypeIDs. Yuck.
```
Please move this to the header and make it a doxygen comment.

roshandathathri reviewed (approved):
LGTM

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -408,6 +408,16 @@ class KATANA_EXPORT PropertyGraph {
 
+  PropertyGraph(
+      katana::GraphTopology&& topo_to_assign,
+      NUMAArray<EntityTypeID> node_entity_type_id,
```
I recommend putting `&&` on moveable objects (or objects to be moved).

scober replied:
Yup, good call. The copy constructor for `NUMAArray` is deleted, so I don't think we can get in too much trouble. But I agree that `&&` is an important readability improvement.
